planner/core/logical_plans.go

```diff
@@ -425,17 +426,59 @@ type DataSource struct {
 type TableGather struct {
 	logicalSchemaProducer
 	Source *DataSource
+	// IsIndexGather marks if this TableGather gathers tuples from an IndexScan.
+	// TableGather and IndexGather are logically equivalent. So that we
+	// only need TableGather for the exploration phase. But in implementation
+	// phase, we need this flag to determine whether to generate PhysicalTableReader
+	// or PhysicalIndexReader.
+	IsIndexGather bool
```

Comment:
Nitpick: how about renaming this `TableGather` to `TiKVSingleGather`? or `SingleGather`?